### PR TITLE
BREAKING CHANGE: In development mode, any new value stored in a tree will be deeply frozen

### LINKED ISSUE #313: Initial state immutability
I am currently exploring immer for a use case of improving an component performance with immutability and angular OnPush change detection. I encountered a strange thing with the auto freezing (which by documentation is as expected) and wanted to know if this is a normal approach or i am pushing to make something work which is not the approach expected by immer.

So i have a complex component state (deeply nested). The sub properties of the object are not created as immutable using the produce but by simple object notation. When produce is called on the state, and the property values are set, immer will only freeze the setter/getter but not the value (which is an object).

I wanted to know if this is something that can be improved by immer or i am tackling this problem wrong and should create every object as immerable or using produce?

### LINKED ISSUE #412: Clarify that the first result from the producer is not immutable unless a change is made
## 🚀 Feature Proposal

As expected, when the producer is first applied to an object (or array) and a change is made, an immutable object results. However, if no such change is made, the result is mutable.

## Motivation

Docs are unclear on this matter. I suggest that it would be better to make the result immutable regardless. But if that doesn't work, it would be good for the docs to be clear on this matter.

With the current implementation, I'm forced to make an arbitrary change to the object to make sure that it is immutable immediately.

## Example

```
    it("Simple object becomes immutable as the produce() call made a change", () => {
        const c = {id: 3};
        const immutable = produce(c, draft => {
            draft.id = 4;
        });
        expect(c.id).eql(3);
        expect(immutable.id).eql(4);
        expect(() => (immutable as any).id = 5).to.throw("");
    });

    it("Simple object is not made immutable as the produce() call made no change", () => {
        const c = {id: 3};
        const immutable = produce(c, draft => {
        });
        (immutable as any).id = 4;
        expect(immutable.id).eql(4);
    });
```

### PR DESCRIPTION
**Pull request by mweststrate**
* [x] Fix #412 #363  #260 #230 
* [x] probably also fix #313, #229 as we will have now side effects during the execution of recipes anyway
* [x] document the side effect with autofreeze enabled!

### CODE REVIEW

**Comment by ianstormtaylor** [comment]
This is awesome, thank you @mweststrate!

File: src/immer.js [hunk 1]
```diff
@@ -299,6 +304,8 @@ export class Immer {
 			// Search new objects for unfinalized drafts. Frozen objects should never contain drafts.
 			else if (isDraftable(value) && !Object.isFrozen(value)) {
 				each(value, finalizeProperty)
+				// for externally incoming object trees, we wan't to make sure they're frozen automatically
```
Comment (aleclarson) [inline src/immer.js]:
```suggestion
				// Freeze new objects when "autoFreeze" is enabled.
```

File: readme.md [hunk 2]
```diff
@@ -548,6 +548,8 @@ isDraft(nextState) // => false
 
 Immer automatically freezes any state trees that are modified using `produce`. This protects against accidental modifications of the state tree outside of a producer. This comes with a performance impact, so it is recommended to disable this option in production. It is by default enabled. By default, it is turned on during local development and turned off in production. Use `setAutoFreeze(true / false)` to explicitly turn this feature on or off.
 
+_⚠️ If auto freezing is enabled, recipes are not entirely side-effect free: Any plain object or array that ends up in the produced result, will be frozen, even when these objects were not frozen before the start of the producer! ⚠️_
```
Comment (aleclarson) [inline readme.md]:
```suggestion
_⚠️ If auto freezing is enabled, recipes are not entirely side-effect free: Any plain object or array that ends up in the produced result will be frozen, even when these objects were not frozen before the start of the producer! ⚠️_
```

File: src/common.js [hunk 3]
```diff
@@ -120,3 +120,12 @@ export function clone(obj) {
...
 	return cloned
 }
+
+export function deepFreeze(obj) {
+	if (!isDraftable(obj)) return
+	if (Object.isFrozen(obj)) return
+	if (isDraft(obj)) return
+	Object.freeze(obj)
+	if (Array.isArray(obj)) obj.forEach(deepFreeze)
+	for (const key in obj) deepFreeze(obj[key])
```
[the remark this answers is not in the document]
Comment (mweststrate) [reply src/common.js]: nope, nice catch!

File: src/immer.js (hunk 1 above)
Comment (mweststrate) [reply src/immer.js]: Removed the comment, didn't really ad anything

File: src/immer.js [hunk 4]
```diff
@@ -87,8 +88,10 @@ export class Immer {
 			return this.processResult(result, scope)
 		} else {
 			result = recipe(base)
-			if (result === undefined) return base
-			return result !== NOTHING ? result : undefined
+			if (result === NOTHING) return undefined
+			if (result === undefined) result = base
+			this.maybeFreeze(result, true)
+			return result
```
[the remark this answers is not in the document]
Comment (mweststrate) [reply src/immer.js]: I think it is just github displaying tabs to be so big. Prettier seems quite happy with it at tabs=2

**Comment by mweststrate** [comment]
Jeez is semantic release so fragile. This didn't trigger a release, can't really see why....

**Comment by mweststrate** [comment]
@aleclarson how to trigger a major release? I followed the manual, but it doesn't recognize the commit?  Any ideas?

**Comment by mweststrate** [comment]
wow, that was burried deep

**Comment by aleclarson** [comment]
You have to put `BREAKING CHANGE` at the end of the commit body

**Comment by mweststrate** [comment]
jeez, I guess my screen contrast is not high enough to deduct that from their readme ... :/

**Pull request merged**